pinot-core/src/main/java/org/apache/pinot/core/operator/combine/MinMaxValueBasedSelectionOrderByCombineOperator.java

```diff
@@ -211,30 +211,51 @@ protected void processSegments() {
           ((AcquireReleaseColumnsSegmentOperator) operator).release();
         }
       }
-      PriorityQueue<Object[]> selectionResult = (PriorityQueue<Object[]>) resultsBlock.getRows();
-      if (selectionResult != null && selectionResult.size() == _numRowsToKeep) {
-        // Segment result has enough rows, update the boundary value
-        assert selectionResult.peek() != null;
-        Comparable segmentBoundaryValue = (Comparable) selectionResult.peek()[0];
-        if (boundaryValue == null) {
-          boundaryValue = segmentBoundaryValue;
-        } else {
-          if (asc) {
-            if (segmentBoundaryValue.compareTo(boundaryValue) < 0) {
-              boundaryValue = segmentBoundaryValue;
-            }
-          } else {
-            if (segmentBoundaryValue.compareTo(boundaryValue) > 0) {
-              boundaryValue = segmentBoundaryValue;
-            }
-          }
-        }
+      Collection<Object[]> rows = resultsBlock.getRows();
+      if (rows != null && rows.size() >= _numRowsToKeep) {
+        boundaryValue = updateBoundaryValue(boundaryValue, extractBoundaryValue(rows), asc);
       }
       threadBoundaryValue = boundaryValue;
       _blockingQueue.offer(resultsBlock);
     }
   }
 
+  private Comparable extractBoundaryValue(Collection<Object[]> rows) {
+    if (rows instanceof PriorityQueue) {
+      PriorityQueue<Object[]> selectionResult = (PriorityQueue<Object[]>) rows;
+      assert selectionResult.peek() != null;
+      // at this point, priority queues are sorted in the inverse order
+      return (Comparable) selectionResult.peek()[0];
+    }
+    List<Object[]> selectionResult;
+    if (rows instanceof List) {
+      selectionResult = (List<Object[]>) rows;
+    } else {
+      LOGGER.warn("Selection result stored in an unexpected data structure of type {}. A copy is needed",
```

Comment:
We are in the paranoid territory here but... most changes I can imagine (for example, adding a new operation that stores rows in a Set) wouldn't break in this situation. But both solutions (falling here or just logging) are not actually that good. What I'm going to do is to attack the cause of the problem and change `SelectionResultsBlock` to only support Lists and PriorityQueues. Therefore is some time in the future we add a new operation datatype, the type system will help us to note that some `SelectionResultsBlock` consumers may assume that rows cannot be of any Collection subclass

Reply:
I don't see how we can enforce that because people can change the code in `SelectionResultsBlock`. What we want to prevent is someone adding `Set` as the rows, and we want it to fail so that people can find the missing code here. Essentially something like:
```
if (rows instanceof PriorityQueue) {
  ...
}
if (rows instanceof List) {
  ...
}
throw new IllegalStateException("Got unexpected rows class: " + rows.getClass());
```

Reply:
I mean, right now rows are always List or PriorityQueue. By specifying that in the constructor, we make it clear. Ideally we should try to move to something closer to my alternative solution with two subclasses, as it was even clearer and type safe in the usage side (ie in case we need to add a `SelectionResultsBlock` whose rows are a set, then the compiler would tell us that we have to do plan how to get the `extractBoundaryValue` in this new class).

But yeah, we can also do what you propose. Also, I think list should also fail if the instance doesn't have a Comparator. Semantically, if there is no comparator, the list doesn't have to be ordered.